File: pokeredus/pokeredus/graph/probabilistic_engine.py
"""
Probabilistic Battle Engine — Monte Carlo simulation over GameState.

Core abstraction: takes a GameState (teams, HP, field, conditions) and
runs multiple rollouts with probabilistic damage sampling to compute
win probabilities and action-value estimates.

Integrates with existing BattleSimulator for SpeciesProfile/MoveEvaluation
lookup and GameState for state management.
"""
from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from pokeredus.graph.game_state import GameState, PokemonState
    from pokeredus.graph.battle_simulator import (
        BattleSimulator, SpeciesProfile, MoveEvaluation, SpeciesMatchupResult,
    )
    from pokeredus.graph.knowledge_graph import KnowledgeGraph
    from pokeredus.graph.matchup_cache_provider import (
        CachedMatchupProvider, MatchupSnapshot,
    )


@dataclass
class SimAction:
    """A possible action in a game state."""
    action_type: str  # 'move' | 'switch'
    target_id: str     # move_id (for 'move') or set_id (for 'switch')
    source_side: str   # 'a' or 'b'

    def __repr__(self) -> str:
        if self.action_type == 'move':
            return f"{self.source_side}:move({self.target_id})"
        return f"{self.source_side}:switch({self.target_id})"


@dataclass
class RolloutResult:
    """Result of a single full-battle Monte Carlo rollout."""
    winner: str          # 'a' | 'b' | 'draw'
    turns: int
    final_hp_a: dict[str, int] = field(default_factory=dict)  # set_id -> HP
    final_hp_b: dict[str, int] = field(default_factory=dict)
    action_log: list[str] = field(default_factory=list)


@dataclass
class ActionStats:
    """Aggregated rollout statistics for one action from a state."""
    action: SimAction
    wins: int = 0
    losses: int = 0
    draws: int = 0
    total_rollouts: int = 0
    avg_turns: float = 0.0

    @property
    def win_rate(self) -> float:
        if self.total_rollouts == 0:
            return 0.0
        return self.wins / self.total_rollouts


@dataclass
class StateEvaluation:
    """Full evaluation of a game state after Monte Carlo sampling."""
    win_probability: float       # 0.0-1.0 from perspective 'a'
    draw_probability: float      # 0.0-1.0
    loss_probability: float      # 0.0-1.0
    actions: list[ActionStats] = field(default_factory=list)   # available actions with their stats
    best_action: Optional[SimAction] = None
    rollout_count: int = 0

    @property
    def perspective_win_pct(self) -> float:
        """Win % formatted for display (0-100)."""
        return self.win_probability * 100.0


class ProbabilisticEngine:
    """
    Monte Carlo battle simulator.

    For a given GameState, runs N rollouts with probabilistic damage
    sampling (picking random values from the 16-roll damage distribution).
    Each rollout simulates the full battle to completion using a
    semi-randomized action selection policy (epsilon-greedy over best moves).

    This is the NUMERIC ABSTRACTION layer — converts complex game state
    into pure probability numbers using existing TTK/matchup infrastructure.

    When a CachedMatchupProvider is set (via ``set_cache_provider()``),
    the engine uses precomputed damage/TTK from the cache instead of
    re-running the full damage formula on every move evaluation.  This
    makes MCTS over large rosters viable in real time.
    """

    def __init__(
        self,
        kg: KnowledgeGraph,
        battle_simulator: BattleSimulator,
        num_rollouts: int = 30,
        max_turns: int = 200,
        seed: int | None = None,
        cache_provider: 'CachedMatchupProvider | None' = None,
    ):
        self.kg = kg
        self.sim = battle_simulator
        self.default_rollouts = num_rollouts
        self.max_turns = max_turns
        self._rng = random.Random(seed)
        self._cache_provider = cache_provider

    # ── Cache provider ──────────────────────────────────────────

    def set_cache_provider(
        self, provider: 'CachedMatchupProvider | None',
    ) -> None:
        """Inject a precomputed cache for fast damage/TTK lookups.

        When set, ``_pick_best_move`` and ``_execute_move`` consult the
        cache before falling back to live damage calculation.
        """
        self._cache_provider = provider

    @property
    def cache_provider(self) -> 'CachedMatchupProvider | None':
        return self._cache_provider

    # ── Public API ──────────────────────────────────────────────────

    def evaluate_state(
        self,
        state: GameState,
        perspective: str = 'a',
        num_rollouts: int | None = None,
    ) -> StateEvaluation:
        """
        Evaluate a game state by running Monte Carlo rollouts.

        Each rollout: clone state, play turns until battle ends or
        max_turns reached, record winner.

        Also evaluates each available top-level action with dedicated rollouts.
        """
        n = num_rollouts or self.default_rollouts

        wins = 0
        losses = 0
        draws = 0

        for _ in range(n):
            clone = state.clone()
            result = self._rollout(clone)
            if result.winner == perspective:
                wins += 1
            elif result.winner == 'draw':
                draws += 1
            else:
                losses += 1

        evaluation = StateEvaluation(
            win_probability=wins / n,
            draw_probability=draws / n,
            loss_probability=losses / n,
            rollout_count=n,
        )

        # Evaluate each available action
        actions = self.get_available_actions(state, perspective)
        action_stats: list[ActionStats] = []

        per_action = max(5, n // max(len(actions), 1))

        for action in actions:
            stats = ActionStats(action=action, total_rollouts=per_action)
            for _ in range(per_action):
                clone = state.clone()
                self._execute_action(clone, action)
                result = self._rollout(clone)
                if result.winner == perspective:
                    stats.wins += 1
                elif result.winner == 'draw':
                    stats.draws += 1
                else:
                    stats.losses += 1
            action_stats.append(stats)

        action_stats.sort(key=lambda s: s.win_rate, reverse=True)
        evaluation.actions = action_stats
        evaluation.best_action = action_stats[0].action if action_stats else None

        return evaluation

    def quick_win_pct(
        self,
        state: GameState,
        perspective: str = 'a',
        num_rollouts: int = 20,
    ) -> float:
        """Fast win% estimate without per-action breakdown."""
        wins = 0
        for _ in range(num_rollouts):
            clone = state.clone()
            result = self._rollout(clone)
            if result.winner == perspective:
                wins += 1
        return wins / num_rollouts

    def get_available_actions(
        self, state: GameState, side: str
    ) -> list[SimAction]:
        """
        Enumerate all possible actions for a side.

        Includes all damaging moves for the active Pokemon and
        switch actions for each alive benched Pokemon.
        """
        active = state.get_active_pokemon(side)
        if not active or active.is_fainted:
            return []

        actions: list[SimAction] = []

        # Moves from the active pokemon's set
        set_obj = self.kg.get_set(active.set_id)
        if set_obj:
            for move_id in set_obj.moves:
                move = self.kg.get_move(move_id)
                if move and not move.is_status:
                    actions.append(SimAction('move', move_id, side))

        # Switches to benched Pokemon
        team = state.team_a if side == 'a' else state.team_b
        active_idx = state.active_a if side == 'a' else state.active_b
        for i, pkmn in enumerate(team):
            if i != active_idx and not pkmn.is_fainted:
                actions.append(SimAction('switch', pkmn.set_id, side))

        return actions

    # ── Rollout Core ────────────────────────────────────────────────

    def _rollout(self, state: GameState) -> RolloutResult:
        """
        Run a single full-battle rollout from the current state.

        Semi-randomized action selection:
        - 70% best available move
        - 20% random move
        - 10% random switch
        """
        for turn in range(self.max_turns):
            is_over, winner = state.is_battle_over()
            if is_over:
                return RolloutResult(winner=winner or 'draw', turns=turn)

            for side in ('a', 'b'):
                active = state.get_active_pokemon(side)
                if not active or active.is_fainted:
                    self._force_switch(state, side)
                    continue

                roll = self._rng.random()
                opponent = state.get_opponent(side)

                if roll < 0.70 and opponent and not opponent.is_fainted:
                    action = self._pick_best_move(state, side)
                elif roll < 0.90:
                    action = self._pick_random_move(state, side)
                else:
                    action = self._pick_random_switch(state, side)

                if action:
                    self._execute_action(state, action)

                is_over, winner = state.is_battle_over()
                if is_over:
                    return RolloutResult(winner=winner or 'draw', turns=turn)

            state.tick()

        # Max turns reached — HP tiebreak
        winner = self._evaluate_hp_tiebreak(state)
        result = RolloutResult(winner=winner, turns=self.max_turns)
        for pkmn in state.team_a:
            result.final_hp_a[pkmn.set_id] = pkmn.current_hp
        for pkmn in state.team_b:
            result.final_hp_b[pkmn.set_id] = pkmn.current_hp
        return result

    # ── Action Selection ────────────────────────────────────────────

    def _pick_best_move(self, state: GameState, side: str) -> Optional[SimAction]:
        """Pick the best damaging move using TTK evaluation.

        If a CachedMatchupProvider is available, we extract the best_move_id
        directly from the cache (O(1) lookup).  Otherwise we fall back to
        iterating all moves through the damage calculator.
        """
        active = state.get_active_pokemon(side)
        opponent = state.get_opponent(side)
        if not active or not opponent:
            return self._pick_random_move(state, side)

        # ── Fast path: cache lookup ─────────────────────────────
        if self._cache_provider and self._cache_provider.is_ready:
            snap = self._cache_provider.lookup(
                active.pokemon_id, opponent.pokemon_id,
            )
            if snap and snap.best_move_id and not snap.is_immune:
                # Verify the cached best move is in the active set
                set_obj = self.kg.get_set(active.set_id)
                if set_obj and snap.best_move_id in set_obj.moves:
                    return SimAction('move', snap.best_move_id, side)

        # ── Slow path: compute via BattleSimulator ──────────────
        our_profile = self.sim.get_profile(active.pokemon_id)
        their_profile = self.sim.get_profile(opponent.pokemon_id)

        if not our_profile or not their_profile:
            return self._pick_random_move(state, side)

        set_obj = self.kg.get_set(active.set_id)
        if not set_obj:
            return self._pick_random_move(state, side)

        best_move_id = None
        best_dmg = 0

        for move_id in set_obj.moves:
            move = self.kg.get_move(move_id)
            if not move or move.is_status:
                continue
            ev = self.sim._evaluate_move(
                move_id, our_profile, their_profile,
                self.sim._get_attribute_modifiers(our_profile),
                self.sim._get_attribute_modifiers(their_profile),
            )
            if ev and not ev.is_immune and ev.avg_damage > best_dmg:
                best_dmg = ev.avg_damage
                best_move_id = move_id

        if best_move_id:
            return SimAction('move', best_move_id, side)
        return self._pick_random_move(state, side)

    def _pick_random_move(self, state: GameState, side: str) -> Optional[SimAction]:
        actions = self.get_available_actions(state, side)
        move_actions = [a for a in actions if a.action_type == 'move']
        if move_actions:
            return self._rng.choice(move_actions)
        return None

    def _pick_random_switch(self, state: GameState, side: str) -> Optional[SimAction]:
        actions = self.get_available_actions(state, side)
        switch_actions = [a for a in actions if a.action_type == 'switch']
        if switch_actions:
            return self._rng.choice(switch_actions)
        return None

    def _force_switch(self, state: GameState, side: str) -> bool:
        """Force switch to first alive benched Pokemon after a faint."""
        team = state.team_a if side == 'a' else state.team_b
        active_idx = state.active_a if side == 'a' else state.active_b

        for i, pkmn in enumerate(team):
            if i != active_idx and not pkmn.is_fainted:
                return state.switch_pokemon(side, i)
        return False

    # ── Action Execution ────────────────────────────────────────────

    def _execute_action(self, state: GameState, action: SimAction) -> None:
        """Execute a SimAction on a game state (mutates in-place)."""
        if action.action_type == 'switch':
            self._execute_switch(state, action)
        elif action.action_type == 'move':
            self._execute_move(state, action)

    def _execute_switch(self, state: GameState, action: SimAction) -> None:
        team = state.team_a if action.source_side == 'a' else state.team_b
        for i, pkmn in enumerate(team):
            if pkmn.set_id == action.target_id and not pkmn.is_fainted:
                state.switch_pokemon(action.source_side, i)
                return

    def _execute_move(self, state: GameState, action: SimAction) -> None:
        """Execute a move: compute probabilistic damage, apply to defender.

        Prefers the cached damage distribution when available (reads
        min_damage/max_damage and samples a roll between them); falls
        back to full damage-calc recomputation on cache miss.
        """
        attacker = state.get_active_pokemon(action.source_side)
        defender = state.get_opponent(action.source_side)

        if not attacker or not defender or defender.is_fainted:
            return

        move = self.kg.get_move(action.target_id)
        if not move or move.is_status:
            return

        # ── Fast path: use cached damage range ──────────────────
        if self._cache_provider and self._cache_provider.is_ready:
            snap = self._cache_provider.lookup(
                attacker.pokemon_id, defender.pokemon_id,
            )
            if snap and not snap.is_immune:
                # Sample damage uniformly between min and max damage
                actual_damage = self._rng.randint(snap.min_damage, snap.max_damage)
                defender.take_damage(max(1, actual_damage))
                attacker.use_move(action.target_id)
                return

        # ── Slow path: full damage-calc recomputation ───────────
        attacker_profile = self.sim.get_profile(attacker.pokemon_id)
        defender_profile = self.sim.get_profile(defender.pokemon_id)

        if not attacker_profile or not defender_profile:
            return

        attacker_mods = self.sim._get_attribute_modifiers(attacker_profile)
        defender_mods = self.sim._get_attribute_modifiers(defender_profile)

        ev = self.sim._evaluate_move(
            action.target_id, attacker_profile, defender_profile,
            attacker_mods, defender_mods,
        )

        if ev is None or ev.is_immune:
            return

        is_physical = move.is_physical
        level = self.sim.calc.level

        if is_physical:
            off_stat = int(max(1, attacker_profile.best_atk * attacker_mods.get('physical_mult', 1.0)))
            def_stat = int(max(1, defender_profile.best_def * defender_mods.get('defense_mult', 1.0)))
        else:
            off_stat = int(max(1, attacker_profile.best_spa * attacker_mods.get('special_mult', 1.0)))
            def_stat = int(max(1, defender_profile.best_spd * defender_mods.get('spdef_mult', 1.0)))

        # Type effectiveness
        type_eff = 1.0
        if hasattr(move, 'type') and move.type:
            try:
                from pokeredus.classes.types import get_effectiveness
                defender_pokemon = self.kg.get_pokemon(defender.pokemon_id)
                if defender_pokemon:
                    type_eff = get_effectiveness(move.type, defender_pokemon.types)
            except ImportError:
                pass

        if type_eff == 0:
            return

        # STAB
        attacker_pokemon = self.kg.get_pokemon(attacker.pokemon_id)
        stab = 1.5 if (
            attacker_pokemon and hasattr(move, 'type') and move.type in attacker_pokemon.types
        ) else 1.0

        # Base damage formula
        base_dmg = int(((2 * level / 5 + 2) * max(1, move.base_power) * off_stat / def_stat) / 50 + 2)
        item_mult = attacker_mods.get('damage_mult', 1.0)

        full_dmg = base_dmg * stab * type_eff * item_mult

        # Probabilistic roll — pick from the 16-roll distribution
        from pokeredus.graph.battle_simulator import DAMAGE_ROLLS
        roll = self._rng.choice(DAMAGE_ROLLS)
        actual_damage = max(1, int(full_dmg * roll))

        defender.take_damage(actual_damage)
        attacker.use_move(action.target_id)

    # ── Tiebreak ────────────────────────────────────────────────────

    def _evaluate_hp_tiebreak(self, state: GameState) -> str:
        """Max turns reached — score based on remaining HP + alive count."""
        hp_a = sum(p.current_hp for p in state.team_a if not p.is_fainted)
        hp_b = sum(p.current_hp for p in state.team_b if not p.is_fainted)
        alive_a = state.count_alive('a')
        alive_b = state.count_alive('b')

        score_a = alive_a * 1000 + hp_a
        score_b = alive_b * 1000 + hp_b

        if score_a > score_b:
            return 'a'
        elif score_b > score_a:
            return 'b'
        return 'draw'

    # ── State Factory ───────────────────────────────────────────────

    def create_state_from_sets(
        self,
        team_a_sets: list[str],
        team_b_sets: list[str],
    ) -> GameState:
        """
        Build a GameState from two lists of set IDs.

        Initializes PokemonState for each set with full HP.
        Sets the first alive Pokemon as active on each side.
        """
        from pokeredus.graph.game_state import GameState, PokemonState

        def build_team(set_ids: list[str]) -> list[PokemonState]:
            team: list[PokemonState] = []
            for sid in set_ids[:6]:
                set_obj = self.kg.get_set(sid)
                if not set_obj:
                    continue
                pokemon = self.kg.get_pokemon(set_obj.pokemon_id)
                if not pokemon:
                    continue

                hp = self._compute_hp(set_obj, pokemon, self.sim.calc.level)

                pstate = PokemonState(
                    pokemon_id=set_obj.pokemon_id,
                    set_id=sid,
                    current_hp=hp,
                    max_hp=hp,
                )
                team.append(pstate)
            return team

        state = GameState(
            team_a=build_team(team_a_sets),
            team_b=build_team(team_b_sets),
        )

        if state.team_a and not state.team_a[0].is_fainted:
            state.team_a[0].is_active = True
            state.active_a = 0
        if state.team_b and not state.team_b[0].is_fainted:
            state.team_b[0].is_active = True
            state.active_b = 0

        return state

    @staticmethod
    def _compute_hp(set_obj, pokemon, level: int) -> int:
        """Compute HP stat for a Pokemon."""
        base = pokemon.base_stats.get('hp', 0)
        iv = set_obj.ivs.get('hp', 31)
        # evs is an EVSpreadClass dataclass, not a dict
        evs = getattr(set_obj, 'evs', None)
        if evs is not None and hasattr(evs, 'get'):
            ev = evs.get('hp')
        else:
            ev = 0
        return int(((2 * base + iv + ev // 4) * level / 100) + level + 10)

File: packages/core/src/battle/probabilistic-engine.ts
import { computeDamage, computeHp, type BattleModifiers } from '@pokeredus/calc';
import type { Move, SetEntry, Species } from '@pokeredus/pack';
import type { KnowledgeGraph } from '../kg/knowledge-graph.js';
import type { SetClass } from '../classes/sets.js';
import { FieldAttribute } from '../classes/attributes.js';
import { GameState, PokemonState } from './game-state.js';
import { BattleSimulator } from './simulator.js';

export interface SimAction {
  action_type: 'move' | 'switch';
  target_id: string;
  source_side: 'a' | 'b';
}

export interface RolloutResult {
  winner: 'a' | 'b' | 'draw';
  turns: number;
  final_hp_a: Record<string, number>;
  final_hp_b: Record<string, number>;
  action_log: string[];
}

export interface ActionStats {
  action: SimAction;
  wins: number;
  losses: number;
  draws: number;
  total_rollouts: number;
  avg_turns: number;
  win_rate: number;
}

export interface StateEvaluation {
  win_probability: number;
  draw_probability: number;
  loss_probability: number;
  actions: ActionStats[];
  best_action: SimAction | null;
  rollout_count: number;
  perspective_win_pct: number;
}

/** ponytail: minimal mulberry32 PRNG for deterministic rollouts */
function mulberry32(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    s += 0x6D2B79F5;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function toPackSet(set: SetClass): SetEntry {
  return {
    id: set.id,
    pokemon_id: set.pokemon_id,
    set_name: set.set_name,
    ability: set.ability,
    item: set.item,
    nature: {
      name: set.nature.name,
      increased_stat: set.nature.increased_stat,
      decreased_stat: set.nature.decreased_stat,
    },
    evs: set.evs.asDict() as unknown as SetEntry['evs'],
    moves: [...set.moves],
    ivs: set.ivs as SetEntry['ivs'],
    role: set.role,
    tera_type: set.tera_type,
  };
}

function toPackSpecies(kg: KnowledgeGraph, pokemonId: string): Species | undefined {
  const p = kg.getPokemon(pokemonId);
  if (!p) return undefined;
  return {
    id: p.id, name: p.name, types: p.types, base_stats: p.base_stats,
    abilities: p.abilities, weight: p.weight, tier: p.tier,
    is_mega: p.is_mega, is_paradox: p.is_paradox, is_legendary: p.is_legendary,
    is_pseudo: p.is_pseudo, api_name: p.api_name, primary_set_id: p.primary_set_id,
  };
}

function mapWeather(w: string | null | undefined): BattleModifiers['weather'] {
  if (!w) return '';
  if (w === 'sun' || w === 'sunny') return 'sunny';
  if (w === 'rain') return 'rain';
  if (w === 'sand' || w === 'sandstorm') return 'sandstorm';
  if (w === 'hail' || w === 'snow') return 'snow';
  return '';
}

function mapTerrain(f: string): BattleModifiers['terrain'] {
  if (f === 'electric_terrain' || f === 'electric') return 'electric';
  if (f === 'grassy_terrain' || f === 'grassy') return 'grassy';
  if (f === 'psychic_terrain' || f === 'psychic') return 'psychic';
  if (f === 'misty_terrain' || f === 'misty') return 'misty';
  return '';
}

function fieldModsFromState(state: GameState): BattleModifiers {
  let terrain: BattleModifiers['terrain'] = '';
  for (const attr of state.field.global_attributes.get({ attribute_type: 'field' })) {
    if (attr instanceof FieldAttribute && FieldAttribute.TERRAIN_TYPES.has(attr.field)) {
      terrain = mapTerrain(attr.field);
      break;
    }
  }
  return {
    weather: mapWeather(state.field.getWeather()),
    terrain,
  };
}

export class ProbabilisticEngine {
  private readonly rng: () => number;

  constructor(
    readonly kg: KnowledgeGraph,
    readonly sim: BattleSimulator,
    readonly defaultRollouts = 30,
    readonly maxTurns = 200,
    seed: number | null = null,
  ) {
    this.rng = seed != null ? mulberry32(seed) : Math.random;
  }

  evaluateState(
    state: GameState,
    perspective: 'a' | 'b' = 'a',
    numRollouts?: number,
  ): StateEvaluation {
    const n = numRollouts ?? this.defaultRollouts;
    let wins = 0;
    let losses = 0;
    let draws = 0;

    for (let i = 0; i < n; i++) {
      const result = this.rollout(state.clone());
      if (result.winner === perspective) wins++;
      else if (result.winner === 'draw') draws++;
      else losses++;
    }

    const evaluation: StateEvaluation = {
      win_probability: wins / n,
      draw_probability: draws / n,
      loss_probability: losses / n,
      actions: [],
      best_action: null,
      rollout_count: n,
      perspective_win_pct: (wins / n) * 100,
    };

    const actions = this.getAvailableActions(state, perspective);
    const perAction = Math.max(5, Math.floor(n / Math.max(actions.length, 1)));

    for (const action of actions) {
      const stats: ActionStats = {
        action, wins: 0, losses: 0, draws: 0,
        total_rollouts: perAction, avg_turns: 0,
        win_rate: 0,
      };
      for (let i = 0; i < perAction; i++) {
        const clone = state.clone();
        this.executeAction(clone, action);
        const result = this.rollout(clone);
        if (result.winner === perspective) stats.wins++;
        else if (result.winner === 'draw') stats.draws++;
        else stats.losses++;
      }
      stats.win_rate = stats.total_rollouts > 0 ? stats.wins / stats.total_rollouts : 0;
      evaluation.actions.push(stats);
    }

    evaluation.actions.sort((a, b) => b.win_rate - a.win_rate);
    evaluation.best_action = evaluation.actions[0]?.action ?? null;
    return evaluation;
  }

  quickWinPct(state: GameState, perspective: 'a' | 'b' = 'a', numRollouts = 20): number {
    let wins = 0;
    for (let i = 0; i < numRollouts; i++) {
      if (this.rollout(state.clone()).winner === perspective) wins++;
    }
    return wins / numRollouts;
  }

  getAvailableActions(state: GameState, side: 'a' | 'b'): SimAction[] {
    const active = state.getActivePokemon(side);
    if (!active || active.isFainted) return [];

    const actions: SimAction[] = [];
    const setObj = this.kg.getSet(active.set_id);
    if (setObj) {
      for (const moveId of setObj.moves) {
        const move = this.kg.getMove(moveId);
        if (move && !move.is_status) {
          actions.push({ action_type: 'move', target_id: moveId, source_side: side });
        }
      }
    }

    const team = side === 'a' ? state.team_a : state.team_b;
    const activeIdx = side === 'a' ? state.active_a : state.active_b;
    for (let i = 0; i < team.length; i++) {
      const pkmn = team[i]!;
      if (i !== activeIdx && !pkmn.isFainted) {
        actions.push({ action_type: 'switch', target_id: pkmn.set_id, source_side: side });
      }
    }
    return actions;
  }

  rollout(state: GameState): RolloutResult {
    for (let turn = 0; turn < this.maxTurns; turn++) {
      const [isOver, winner] = state.isBattleOver();
      if (isOver) {
        return { winner: (winner ?? 'draw') as 'a' | 'b' | 'draw', turns: turn, final_hp_a: {}, final_hp_b: {}, action_log: [] };
      }

      for (const side of ['a', 'b'] as const) {
        const active = state.getActivePokemon(side);
        if (!active || active.isFainted) {
          this.forceSwitch(state, side);
          continue;
        }

        const roll = this.rng();
        const opponent = state.getOpponent(side);
        let action: SimAction | null = null;

        if (roll < 0.70 && opponent && !opponent.isFainted) {
          action = this.pickBestMove(state, side);
        } else if (roll < 0.90) {
          action = this.pickRandomMove(state, side);
        } else {
          action = this.pickRandomSwitch(state, side);
        }

        if (action) this.executeAction(state, action);

        const [over, w] = state.isBattleOver();
        if (over) {
          return { winner: (w ?? 'draw') as 'a' | 'b' | 'draw', turns: turn, final_hp_a: {}, final_hp_b: {}, action_log: [] };
        }
      }
      state.tick();
    }

    const winner = this.evaluateHpTiebreak(state);
    const result: RolloutResult = {
      winner, turns: this.maxTurns, action_log: [],
      final_hp_a: {}, final_hp_b: {},
    };
    for (const p of state.team_a) result.final_hp_a[p.set_id] = p.current_hp;
    for (const p of state.team_b) result.final_hp_b[p.set_id] = p.current_hp;
    return result;
  }

  pickBestMove(state: GameState, side: 'a' | 'b'): SimAction | null {
    const active = state.getActivePokemon(side);
    const opponent = state.getOpponent(side);
    if (!active || !opponent) return this.pickRandomMove(state, side);

    const ourProfile = this.sim.getProfile(active.pokemon_id);
    const theirProfile = this.sim.getProfile(opponent.pokemon_id);
    const setObj = this.kg.getSet(active.set_id);
    if (!setObj) return this.pickRandomMove(state, side);

    const ourMods = this.sim.getAttributeModifiers(ourProfile);
    const theirMods = this.sim.getAttributeModifiers(theirProfile);
    const fieldMods = fieldModsFromState(state);

    let bestMoveId: string | null = null;
    let bestDmg = 0;

    for (const moveId of setObj.moves) {
      const move = this.kg.getMove(moveId);
      if (!move || move.is_status) continue;
      const ev = this.sim.evaluateMove(moveId, ourProfile, theirProfile, ourMods, theirMods, fieldMods);
      if (ev && !ev.is_immune && ev.avg_damage > bestDmg) {
        bestDmg = ev.avg_damage;
        bestMoveId = moveId;
      }
    }

    return bestMoveId
      ? { action_type: 'move', target_id: bestMoveId, source_side: side }
      : this.pickRandomMove(state, side);
  }

  pickRandomMove(state: GameState, side: 'a' | 'b'): SimAction | null {
    const moves = this.getAvailableActions(state, side).filter((a) => a.action_type === 'move');
    if (!moves.length) return null;
    return moves[Math.floor(this.rng() * moves.length)]!;
  }

  pickRandomSwitch(state: GameState, side: 'a' | 'b'): SimAction | null {
    const switches = this.getAvailableActions(state, side).filter((a) => a.action_type === 'switch');
    if (!switches.length) return null;
    return switches[Math.floor(this.rng() * switches.length)]!;
  }

  forceSwitch(state: GameState, side: 'a' | 'b'): boolean {
    const team = side === 'a' ? state.team_a : state.team_b;
    const activeIdx = side === 'a' ? state.active_a : state.active_b;
    for (let i = 0; i < team.length; i++) {
      if (i !== activeIdx && !team[i]!.isFainted) {
        return state.switchPokemon(side, i);
      }
    }
    return false;
  }

  executeAction(state: GameState, action: SimAction): void {
    if (action.action_type === 'switch') this.executeSwitch(state, action);
    else if (action.action_type === 'move') this.executeMove(state, action);
  }

  executeSwitch(state: GameState, action: SimAction): void {
    const team = action.source_side === 'a' ? state.team_a : state.team_b;
    for (let i = 0; i < team.length; i++) {
      const pkmn = team[i]!;
      if (pkmn.set_id === action.target_id && !pkmn.isFainted) {
        state.switchPokemon(action.source_side, i);
        return;
      }
    }
  }

  executeMove(state: GameState, action: SimAction): void {
    const attacker = state.getActivePokemon(action.source_side);
    const defender = state.getOpponent(action.source_side);
    if (!attacker || !defender || defender.isFainted) return;

    const move = this.kg.getMove(action.target_id);
    if (!move || move.is_status) return;

    const attackerSet = this.kg.getSet(attacker.set_id);
    const defenderSet = this.kg.getSet(defender.set_id);
    const atkSp = toPackSpecies(this.kg, attacker.pokemon_id);
    const defSp = toPackSpecies(this.kg, defender.pokemon_id);
    if (!attackerSet || !defenderSet || !atkSp || !defSp) return;

    const fieldMods = fieldModsFromState(state);
    const result = computeDamage(
      toPackSet(attackerSet), toPackSet(defenderSet),
      {
        id: move.id, name: move.name, type: move.type,
        category: move.category as Move['category'],
        base_power: move.base_power,
        accuracy: move.accuracy as Move['accuracy'],
        priority: move.priority, pp: move.pp, target: move.target,
        flags: move.flags, secondary_effects: move.secondary_effects,
      },
      atkSp, defSp, this.sim.level, fieldMods,
    );

    if (result.is_immune) return;

    const { min_damage: minDmg, max_damage: maxDmg } = result;
    const actualDamage = maxDmg <= minDmg
      ? maxDmg
      : minDmg + Math.floor(this.rng() * (maxDmg - minDmg + 1));
    defender.takeDamage(Math.max(1, actualDamage));
    attacker.useMove(action.target_id);
  }

  evaluateHpTiebreak(state: GameState): 'a' | 'b' | 'draw' {
    const hpA = state.team_a.filter((p) => !p.isFainted).reduce((s, p) => s + p.current_hp, 0);
    const hpB = state.team_b.filter((p) => !p.isFainted).reduce((s, p) => s + p.current_hp, 0);
    const scoreA = state.countAlive('a') * 1000 + hpA;
    const scoreB = state.countAlive('b') * 1000 + hpB;
    if (scoreA > scoreB) return 'a';
    if (scoreB > scoreA) return 'b';
    return 'draw';
  }

  createStateFromSets(teamASetIds: string[], teamBSetIds: string[]): GameState {
    const buildTeam = (setIds: string[]): PokemonState[] => {
      const team: PokemonState[] = [];
      for (const sid of setIds.slice(0, 6)) {
        const setObj = this.kg.getSet(sid);
        if (!setObj) continue;
        const sp = toPackSpecies(this.kg, setObj.pokemon_id);
        if (!sp) continue;
        const hp = computeHp(sp, toPackSet(setObj), this.sim.level);
        const ps = new PokemonState(setObj.pokemon_id, sid);
        ps.current_hp = hp;
        ps.max_hp = hp;
        team.push(ps);
      }
      return team;
    };

    const state = new GameState();
    state.team_a = buildTeam(teamASetIds);
    state.team_b = buildTeam(teamBSetIds);

    if (state.team_a[0] && !state.team_a[0].isFainted) {
      state.team_a[0].is_active = true;
      state.active_a = 0;
    }
    if (state.team_b[0] && !state.team_b[0].isFainted) {
      state.team_b[0].is_active = true;
      state.active_b = 0;
    }
    return state;
  }
}
